src/core/ext/filters/client_channel/lb_policy/rls/rls.cc

```diff
@@ -1116,40 +1116,45 @@ LoadBalancingPolicy::PickResult RlsLb::Picker::Pick(PickArgs args) {
 RlsLb::Cache::Entry::BackoffTimer::BackoffTimer(RefCountedPtr<Entry> entry,
                                                 Timestamp backoff_time)
     : entry_(std::move(entry)) {
-  GRPC_CLOSURE_INIT(&backoff_timer_callback_, OnBackoffTimer, this, nullptr);
-  Ref(DEBUG_LOCATION, "BackoffTimer").release();
-  grpc_timer_init(&backoff_timer_, backoff_time, &backoff_timer_callback_);
+  backoff_timer_task_handle_ =
+      entry_->lb_policy_->channel_control_helper()->GetEventEngine()->RunAfter(
+          backoff_time - Timestamp::Now(),
+          [self = Ref(DEBUG_LOCATION, "BackoffTimer")] {
+            self->OnBackoffTimer();
+          });
 }
 
 void RlsLb::Cache::Entry::BackoffTimer::Orphan() {
-  if (armed_) {
-    armed_ = false;
-    grpc_timer_cancel(&backoff_timer_);
+  if (backoff_timer_task_handle_.has_value()) {
+    if (entry_->lb_policy_->channel_control_helper()->GetEventEngine()->Cancel(
+            *backoff_timer_task_handle_)) {
+      if (GRPC_TRACE_FLAG_ENABLED(grpc_lb_rls_trace)) {
+        gpr_log(
+            GPR_INFO, "[rlslb %p] cache entry=%p %s, backoff timer canceled",
+            entry_->lb_policy_.get(), entry_.get(),
+            entry_->is_shutdown_ ? "(shut down)"
+                                 : entry_->lru_iterator_->ToString().c_str());
+      }
+    }
+    backoff_timer_task_handle_.reset();
   }
   Unref(DEBUG_LOCATION, "Orphan");
 }
 
-void RlsLb::Cache::Entry::BackoffTimer::OnBackoffTimer(
-    void* arg, grpc_error_handle /*error*/) {
-  auto* self = static_cast<BackoffTimer*>(arg);
-  self->entry_->lb_policy_->work_serializer()->Run(
-      [self]() {
-        RefCountedPtr<BackoffTimer> backoff_timer(self);
+void RlsLb::Cache::Entry::BackoffTimer::OnBackoffTimer() {
+  entry_->lb_policy_->work_serializer()->Run(
+      [self = Ref(DEBUG_LOCATION, "BackoffTimer-WorkSerializer")]() {
         {
           MutexLock lock(&self->entry_->lb_policy_->mu_);
+          self->backoff_timer_task_handle_.reset();
           if (GRPC_TRACE_FLAG_ENABLED(grpc_lb_rls_trace)) {
             gpr_log(GPR_INFO,
-                    "[rlslb %p] cache entry=%p %s, armed_=%d: "
-                    "backoff timer fired",
+                    "[rlslb %p] cache entry=%p %s, backoff timer fired",
                     self->entry_->lb_policy_.get(), self->entry_.get(),
                     self->entry_->is_shutdown_
                         ? "(shut down)"
-                        : self->entry_->lru_iterator_->ToString().c_str(),
-                    self->armed_);
+                        : self->entry_->lru_iterator_->ToString().c_str());
           }
-          bool cancelled = !self->armed_;
-          self->armed_ = false;
-          if (cancelled) return;
```

Comment:
Do we need to honor the early return here if we have already cancelled the timer (through `Orphan()`)? We could check `if (self->backoff_timer_task_handle_.has_value())` provided if it hasn't been reset above.

Reply:
Good question. Previously `OnBackoffTimer` would be called whether the timer was cancelled or not, hence the need to check the condition inside the callback. Now, `OnBackoffTimer` is only called if the timer has _not_ been cancelled. The logic for what to do on cancellation was moved to line 1131. So I think it's correct as is.

Reply:
I see what you're saying now, the `cancelled` value in the callback is conflated with the concept of being orphaned. That callback had different behavior when orphaned, regardless of the timer cancellation status. Fixed.